Test `test_zookeeper_connection_log/test.py::test_zookeeper_connection_log` is flaky

https://play.clickhouse.com/play?user=play&run=1#V0lUSAogICAgOTAgQVMgaW50ZXJ2YWxfZGF5cwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkYXksCiAgICBjb3VudCgpIEFTIGZhaWx1cmVzLAogICAgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlcikgQVMgcHJzLAogICAgYW55KHJlcG9ydF91cmwpIEFTIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZQogICAgQU5EIHRlc3RfbmFtZSA9ICd0ZXN0X3pvb2tlZXBlcl9jb25uZWN0aW9uX2xvZy90ZXN0LnB5Ojp0ZXN0X3pvb2tlZXBlcl9jb25uZWN0aW9uX2xvZycKICAgIC0tIEFORCBjaGVja19uYW1lID0gJ0ludGVncmF0aW9uIHRlc3RzIChhcm1fYmluYXJ5LCBkaXN0cmlidXRlZCBwbGFuLCA0LzQpJwogICAgQU5EIHRlc3Rfc3RhdHVzIElOICgnRkFJTCcsICdFUlJPUicpCiAgICBBTkQgKChoZWFkX3JlZiA9ICdtYXN0ZXInIEFORCBwdWxsX3JlcXVlc3RfbnVtYmVyID0gMCkgT1IgcHVsbF9yZXF1ZXN0X251bWJlciAhPSAwKQpHUk9VUCBCWSBkYXkKT1JERVIgQlkgZGF5IERFU0MK

Example:
```
File: test_zookeeper_connection_log/test.py:142 - in test_zookeeper_connection_log
    assert (
E   assert node1	Connected	default	zoo1	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Connected	default	zoo1	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Connected	zk_conn_log_test_2	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Connected	zk_conn_log_test_3	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Disconnected	default	zoo1	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Connected	default	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Disconnected	zk_conn_log_test_2	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Connected	zk_conn_log_test_2	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Disconnected	zk_conn_log_test_3	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Removed from config\nnode1	Connected	zk_conn_log_test_4	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization == node1	Connected	default	zoo1	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Connected	zk_conn_log_test_2	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Connected	zk_conn_log_test_3	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Disconnected	default	zoo1	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Connected	default	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Disconnected	zk_conn_log_test_2	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Connected	zk_conn_log_test_2	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Disconnected	zk_conn_log_test_3	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Removed from config\nnode1	Connected	zk_conn_log_test_4	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization
E    +  where node1	Connected	default	zoo1	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Connected	default	zoo1	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Connected	zk_conn_log_test_2	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Connected	zk_conn_log_test_3	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization\nnode1	Disconnected	default	zoo1	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Connected	default	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Disconnected	zk_conn_log_test_2	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Connected	zk_conn_log_test_2	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Config changed\nnode1	Disconnected	zk_conn_log_test_3	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Removed from config\nnode1	Connected	zk_conn_log_test_4	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']	Initialization = TSV("node1\tConnected\tdefault\tzoo1\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\nnode1\tConnected\tdefault\tzoo1\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\nnode1\tConnected\tzk_conn_log_test_2\tzoo2\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\nnode1\tConnected\tzk_conn_log_test_3\tzoo3\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\nnode1\tDisconnected\tdefault\tzoo1\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tConfig changed\nnode1\tConnected\tdefault\tzoo2\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tConfig changed\nnode1\tDisconnected\tzk_conn_log_test_2\tzoo2\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tConfig changed\nnode1\tConnected\tzk_conn_log_test_2\tzoo3\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tConfig changed\nnode1\tDisconnected\tzk_conn_log_test_3\tzoo3\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tRemoved from config\nnode1\tConnected\tzk_conn_log_test_4\tzoo2\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\n")
E    +    where "node1\tConnected\tdefault\tzoo1\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\nnode1\tConnected\tdefault\tzoo1\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\nnode1\tConnected\tzk_conn_log_test_2\tzoo2\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\nnode1\tConnected\tzk_conn_log_test_3\tzoo3\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\nnode1\tDisconnected\tdefault\tzoo1\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tConfig changed\nnode1\tConnected\tdefault\tzoo2\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tConfig changed\nnode1\tDisconnected\tzk_conn_log_test_2\tzoo2\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tConfig changed\nnode1\tConnected\tzk_conn_log_test_2\tzoo3\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tConfig changed\nnode1\tDisconnected\tzk_conn_log_test_3\tzoo3\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tRemoved from config\nnode1\tConnected\tzk_conn_log_test_4\tzoo2\t2181\t0\t0\t['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES','CREATE_WITH_STATS']\tInitialization\n" = query("SELECT hostname, type, name, host, port, index, keeper_api_version, enabled_feature_flags, reason\n                               FROM system.zookeeper_connection_log\n                               WHERE event_time_microseconds >= '2025-11-21 03:27:01.399'\n                               ORDER BY event_time_microseconds")
E    +      where query = <helpers.cluster.ClickHouseInstance object at 0xff12ae2bcd60>.query
```

--- a/tests/integration/test_zookeeper_connection_log/test.py
+++ b/tests/integration/test_zookeeper_connection_log/test.py
@@ -108,7 +108,7 @@ def test_zookeeper_connection_log(started_cluster):
                 logging.debug(
                     f"Checking for 8 rows in zookeeper_connection_log, got: {res}"
                 )
-                return res == "8\n"
+                return int(res) >= 8
 
             node1.query_with_retry(
                 f"SELECT count() FROM system.zookeeper_connection_log WHERE event_time_microseconds >= '{test_start_time}'",
@@ -123,10 +123,31 @@ def test_zookeeper_connection_log(started_cluster):
 
             logging.debug(
                 node1.query(
-                    """SELECT event_time_microseconds, hostname, type, name, host, port, index, keeper_api_version, enabled_feature_flags, reason
+                    """SELECT event_time_microseconds, client_id, hostname, type, name, host, port, index, keeper_api_version, enabled_feature_flags, reason
                                FROM system.zookeeper_connection_log  ORDER BY event_time_microseconds"""
                 )
             )
+            logging.debug(
+                node1.query(
+                    f"""SELECT hostname, type, name, host, port, reason, count() AS cnt,
+                               groupArray(client_id) AS client_ids
+                           FROM system.zookeeper_connection_log
+                          WHERE event_time_microseconds >= '{test_start_time}'
+                          GROUP BY hostname, type, name, host, port, reason
+                          ORDER BY min(event_time_microseconds)"""
+                )
+            )
+            normalized_rows = node1.query(
+                f"""SELECT hostname, type, name, host, port, index, keeper_api_version,
+                               arrayFilter(x -> has(enabled_feature_flags, x),
+                                           ['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES']) AS enabled_feature_flags,
+                               reason
+                          FROM system.zookeeper_connection_log
+                         WHERE event_time_microseconds >= '{test_start_time}'
+                         ORDER BY event_time_microseconds
+                         LIMIT 1 BY hostname, type, name, host, port, reason"""
+            )
+            logging.debug("Normalized rows used in comparison:\n" + normalized_rows)
             expected = TSV(
                 """node1	Connected	default	zoo1	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES']	Initialization
 node1	Connected	zk_conn_log_test_2	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES']	Initialization
@@ -139,17 +160,7 @@ node1	Disconnected	zk_conn_log_test_3	zoo3	2181	0	0	['FILTERED_LIST','MULTI_READ
 node1	Connected	zk_conn_log_test_4	zoo2	2181	0	0	['FILTERED_LIST','MULTI_READ','CHECK_NOT_EXISTS','CREATE_IF_NOT_EXISTS','REMOVE_RECURSIVE','MULTI_WATCHES']	Initialization"""
             )
 
-            assert (
-                TSV(
-                    node1.query(
-                        f"""SELECT hostname, type, name, host, port, index, keeper_api_version, enabled_feature_flags, reason
-                               FROM system.zookeeper_connection_log
-                               WHERE event_time_microseconds >= '{test_start_time}'
-                               ORDER BY event_time_microseconds"""
-                    )
-                )
-                == expected
-            )
+            assert TSV(normalized_rows) == expected
             assert (
                 int(
                     node1.query(

